### PR TITLE
feat: handle multiple ResponseInfo obj in Query response

### LINKED ISSUE #605: Query response may contain ResponseInfo[1..*]
According to the [CORE-28 spec clause](https://docs.oasis-open-projects.org/oslc-op/core/v3.0/os/oslc-core.html#cc-the-subject-resource-uri-of-the-oslc-responseinfo-resource-representation-must-be-the-http-request-uri-or-the-uri-from-subsequent-redirects.-the-response-representation-may-also-include-properties-from-subject-resources-different-from-the-one-identified-by-the-request-uri.x), at least one ResponseInfo object must have the same URI as the query URI or some redirect URI. This could be used to find the "main" ResposeInfo object.

### PR DESCRIPTION
**Pull request by berezovskyi**
## Description

Allows Query responses to be paged even in presence of multiple ResponseInfo objects.

## Checklist

- [x] This PR adds an entry to the CHANGELOG. _See https://keepachangelog.com/en/1.0.0/ for instructions. Minor edits are exempt._
- [x] This PR was tested on at least one Lyo OSLC server (e.g. manual workflow on [Lyo Sample](https://github.com/OSLC/lyo-samples/actions/workflows/maven-smoke-manual.yml) and [OSLC RefImpl](https://github.com/oslc-op/refimpl/actions/workflows/maven-acceptance-manual.yml)) or adds unit/integration tests.
- [x] This PR does NOT break the API

## Issues

Closes #605

See:

- https://forum.open-services.net/t/cannot-obtain-nextpage-url-for-ewm-query-multiple-responseinfo-tags-in-the-query-result/825
- https://github.com/eclipse/lyo/issues/605
- https://github.com/OSLC/oslc4net/issues/185

### CODE REVIEW

File: client/oslc-client/src/main/java/org/eclipse/lyo/client/query/OslcQueryResult.java
Comment (Jad-el-khoury) [inline client/oslc-client/src/main/java/org/eclipse/lyo/client/query/OslcQueryResult.java:135]: would it make more sense to replace "equals" with startsWith?

Comment (berezovskyi) [reply client/oslc-client/src/main/java/org/eclipse/lyo/client/query/OslcQueryResult.java:135]: Well, there is no strict rule on the ordering of query params, so the substring check does not have a strong meaning. I was thinking to just cut off all the query params and compare the URLs with all params stripped but I wanted to hear from Ana how her testing goes before altering the URIs.

Comment (Jad-el-khoury) [reply client/oslc-client/src/main/java/org/eclipse/lyo/client/query/OslcQueryResult.java:135]: True.

Comment (anamrosu) [reply client/oslc-client/src/main/java/org/eclipse/lyo/client/query/OslcQueryResult.java:135]: Hi Andrew, don't know if you saw it already, but please check my feedback on [the forum](https://forum.open-services.net/t/cannot-obtain-nextpage-url-for-ewm-query-multiple-responseinfo-tags-in-the-query-result/825/8?u=anamr).

Thanks,
Ana

Comment (berezovskyi) [reply client/oslc-client/src/main/java/org/eclipse/lyo/client/query/OslcQueryResult.java:135]: @anamrosu, thanks! Sad to hear a direct check did not succeed.

I am thinking of keeping the current checks and adding a substring check as a fallback. 

Unfortunately, I am really busy this week and will not be able to work on the PR (I do not work on Lyo during my day job). Also, we are now in the release freeze period after Jad published the Lyo 6 release candidate. We won't merge any PRs except critical bugfixes until next week.

@anamrosu if you are up for it, I would be glad if you fork the repo, update my code and submit a new PR. Thank you for helping us improve Lyo!

Comment (berezovskyi) [reply client/oslc-client/src/main/java/org/eclipse/lyo/client/query/OslcQueryResult.java]: @anamrosu just updated the logic, please check.

Comment (anamrosu) [reply client/oslc-client/src/main/java/org/eclipse/lyo/client/query/OslcQueryResult.java]: Hi Andrew, sorry for not giving you any updates, I was also pretty busy the past weeks. I will be on vacation starting tomorrow for 2 weeks, so I will test your changes after I come back. Thanks!

**Comment by berezovskyi** [comment]
Also fixed in https://github.com/OSLC/oslc4net/pull/203

File: client/oslc-client/src/main/java/org/eclipse/lyo/client/query/OslcQueryResult.java
Comment (anamrosu) [reply client/oslc-client/src/main/java/org/eclipse/lyo/client/query/OslcQueryResult.java]: Hi Andrew, I tested the changes a few days ago and it was still not working for us. So, I thought about an alternative solution which, in my opinion, should work better. I also tested it in our environment and it works.
In this solution I'm trying to directly look for the responseInfo containg "nextPage" in it. There should be exactly one responseInfo matching this case. I created a new pull request with my proposal, please have a look: https://github.com/eclipse-lyo/lyo/pull/646

Comment (berezovskyi) [reply client/oslc-client/src/main/java/org/eclipse/lyo/client/query/OslcQueryResult.java]: Thanks @anamrosu. I updated my PR. Now the logic goes through the following:

- exactly one ResponseInfo
- otherwise, exactly one ResponseInfo with the nextPage (added in the update)
- otherwise, exactly one ResponseInfo with the exact URI match
- otherwise, exactly one ResponseInfo with the prefix URI match

I could not use your PR directly because:

- you did not sign Eclipse CLA yet: https://api.eclipse.org/git/eca/status/gh/eclipse-lyo/lyo/646 - please create an account with Eclipse and sign the CLA.
- there is a risk that multiple response infos will have the next page (paging is part of OSLC Core, not Query - any valid OSLC Core response may have paging)
- the last page will not have the nextPage but we still need to locate the response info because it contains extra info than just the next page (totalCount, for example - I exposed it in the "sister" PR for oslc4net https://github.com/OSLC/oslc4net/pull/203)
- I wanted to add some logging to help with future improvements - your one-shot iteration was a bit hard to use for that.

**Comment by berezovskyi** [comment]
@Jad-el-khoury @anamrosu please approve the PR if you think we are ready to merge and release this as SNAPSHOT

**Review by Jad-el-khoury — APPROVED** [review]

**Comment by Jad-el-khoury** [comment]
The logic looks ok to me, assuming it is tested.
I think we should raise this issue to the OP, since the specs makes the response quite unstable.

**Comment by berezovskyi** [comment]
@Jad-el-khoury thanks, there is a unit test inside using a real OSLC query response I got from IBM CLM Nordic instance we have access to.

@anamrosu please check if 7.0.0-SNAPSHOT works for you as expected. If not, please comment/reopen #605 and we will investigate further.

File: client/oslc-client/src/main/java/org/eclipse/lyo/client/query/OslcQueryResult.java
Comment (anamrosu) [reply client/oslc-client/src/main/java/org/eclipse/lyo/client/query/OslcQueryResult.java]: Hi @berezovskyi, thanks for the detailed explanation and for taking my solution into account when merging the pull request. Do you have an estimate date when these changes will be officially released?

**Comment by anamrosu** [comment]
I tested just now and for us it works well.

**Comment by berezovskyi** [comment]
@anamrosu great to hear!


The release of Lyo 7 is tied to the upgrade to Jena 5. We are on Jena 4 now and no versions below Jena 5 are receiving security updates any more. If we get any help with it, we can release it sooner. We have a PR open to upgrade to Jena 4.10 that needs testing and then we can do the remaining updates needed for Jena 5.

I can maybe release a 7.0.0-alpha1 for you if it will help you and if I can find some time to do it. @Jad-el-khoury what do you think about doing an alpha for this patch and Jena 4.10?

**Comment by anamrosu** [comment]
Hi @berezovskyi, an Alpha version would be helpful, thanks!

**Comment by berezovskyi** [comment]
@anamrosu https://forum.open-services.net/t/new-versions-of-lyo-and-oslc4net-released/839 there are two alphas for you to try

**Pull request merged**
